Objects/longobject.c

```diff
@@ -3220,6 +3220,8 @@ x_sub(PyLongObject *a, PyLongObject *b)
             ;
         if (i < 0)
             return (PyLongObject *)PyLong_FromLong(0);
+        if (i == 0)
+            return (PyLongObject *)PyLong_FromLong((sdigit)a->ob_digit[0] - (sdigit)b->ob_digit[0]);
```

Comment:
Can we replace it with this?
```c
while (--i > 0 && a->ob_digit[i] == b->ob_digit[i])
    ;
if (i == 0)
    return (PyLongObject *)PyLong_FromLong((sdigit)a->ob_digit[0] - (sdigit)b->ob_digit[0]);
```

Reply:
emm, do you mean `if (i <= 0)` ?

Reply:
I also changed condition in `while` loop, we don't need to compare last digit, right?

Reply:
But yes it could be `if (i <= 0)` :-)

Reply:
Please revert this change and put it in a separated PR. It looks unrelated.

Reply:
Oh, ok.

Reply:
@vstinner 

I think this change is related to this PR because it handles small ints too, don't it?

Reply:
Is it more reasonable to make a new PR for this optimization after this PR merge?

Reply:
> I think this change is related to this PR because it handles small ints too, don't it?

This PR makes possible to return a small int (singleton) when previously it was possible.

This specific change is an optimization, it's unrelated.

> Is it more reasonable to make a new PR for this optimization after this PR merge?

Yes, I would prefer to see a separated PR just for this optimization, once this PR is merged.

Apart of this optimization, the PR LGTM.

Reply:
> This PR makes possible to return a small int (singleton) when previously it was possible.

Why is this desirable, if it makes performance worse? I think that the main purpose of using these singletons is performance. We don't guarantee that these singletons are always used.

Reply:
> Why is this desirable, if it makes performance worse?

It seems like the singleton change + the optimization basically have no significant impact on performances.

I prefer to have 2 separated commits because it makes the purpose of each change more obvious, it is easier to revert in case of regression, etc.